### PR TITLE
Quorum replication is now feature complete. Please take a look.

### PR DESCRIPTION
**Pull request by kspirov**
Merge on the top of the refactoring of Michael.

* StreamsView is not aware any longer about the REPLEX and CHAIN replication. The factory method accesses the ReplicationMode enum.
* In the same manner I removed the switch for hardcoded ReplicationModes in AbstractReplicationView
* I also eliminated the flags 'override' and 'replexOverride' in TokenRequest and their reference in many place. Removed some related code in StreamView. Actually these flags were deprecated by you and you don't use them any longer.
* Writer and fillHole send write request to all servers, but wait only n/2+1 servers respond.
* Fixed some obvious bugs in AbstractCorfuTest (2 minor thread safety isssues and wrong position of one as() from assertJs)
* Inherited some tests for the quorum replication
* Quorum replication view - Reader, Writer and fillHole send write request to all servers, but wait only n/2+1 servers respond.
* The logic how to use the futures is encapsulated in a new QuorumReplicationFuture which is separately testable. This future does not implements any locking itself, it  just delegates the locking to the leafs.  Also no any threads are added manually, just the internally CompletableFuture leafs use the ManagedBlocker in their ForkJoinPool.  I believe this approach is novel, at least I don't see anything similar in the public code repos. This helps us not only with the testability, but also we also keep the replication view readable.

Non-steady states: In case of the server failures, this approach by concept will be more solid, as it will continue to read and write successfully (the chain implementation would block during the write).
What needs more attention (and what is the primary reason this implementation is "draft") - the behaviour when the sequencer fails needs more attention (+ dedicated tests).  This is exactly what I am doing now.

### CODE REVIEW

File: runtime/src/main/java/org/corfudb/runtime/view/QuorumFutureFactory.java [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+                                numCompleteFutures++;
+                            }
+                        }
+                    }
+                    int quorum = futures.length/2+1;
+                    if (numCompleteFutures>=futuresToWait) { // normal exit, quorum
+                        done = true;
+                        return futures[lastCompleteFutureIndex].get();
+                    }
+                    if (numAbnormalExists>=quorum) { // abnormal exit, canceled or failed futures
```
Comment (kjames88) [inline runtime/src/main/java/org/corfudb/runtime/view/QuorumFutureFactory.java]: If futuresToWait != futures.length/2+1 is it correct to exit when numAbnormalExists>=quorum or should it wait until futuresToWait can no longer be satisfied?

File: runtime/src/main/java/org/corfudb/runtime/view/QuorumReplicationView.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+     */
+    @Override
+    public LogData read(long address) {
+        int numUnits = getLayout().getSegmentLength(address);
+        log.trace("Read[{}]: chain {}/{}", address, numUnits, numUnits);
+        CompletableFuture<ReadResponse>[] futures = new CompletableFuture[numUnits];
+            for (int i=0; i<numUnits; i++) {
+                futures[i]=getLayout().getLogUnitClient(address, i).read(address);
+        }
+        ReadResponse readResponse = CFUtils.getUninterruptibly(QuorumFutureFactory.getQuorumFuture(futures));
```
Comment (kjames88) [inline runtime/src/main/java/org/corfudb/runtime/view/QuorumReplicationView.java]: What if we get conflicting responses?  It looks like the QuorumFuture will just return the latest.

File: runtime/src/main/java/org/corfudb/runtime/view/QuorumFutureFactory.java (hunk 1 above)
Comment (dahliamalkhi) [inline runtime/src/main/java/org/corfudb/runtime/view/QuorumFutureFactory.java]: I think this is fine. The assumption is that every pair of quorums intersect, that is, there is no distinction between write- and read- quorums. Hence, if a quorum failed, there is no hope of getting a quorum of responses.

File: runtime/src/main/java/org/corfudb/runtime/view/QuorumReplicationView.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    /**
+     * Read the given object from an address, using the replication method given.
+     *
+     * @param address The address to read from.
+     * @return The result of the read.
+     */
+    @Override
+    public LogData read(long address) {
+        int numUnits = getLayout().getSegmentLength(address);
+        log.trace("Read[{}]: chain {}/{}", address, numUnits, numUnits);
```
Comment (dahliamalkhi) [inline runtime/src/main/java/org/corfudb/runtime/view/QuorumReplicationView.java]: why say chain here?

File: runtime/src/main/java/org/corfudb/runtime/view/QuorumReplicationView.java (hunk 2 above)
Comment (dahliamalkhi) [inline runtime/src/main/java/org/corfudb/runtime/view/QuorumReplicationView.java]: Agreed. Here, we would need to address all the interesting cases.

As a first cut, I would simply return "unwritten" for now. Just to get some tests rolling.

In the future:
1. If less than a quorum return the written result, try to write-back to complete the quorum 
2. if two responses contain different values (one would be a hole), then do a 2-phase "Paxos" consensus protocol

File: runtime/src/main/java/org/corfudb/runtime/view/QuorumReplicationView.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+     *
+     * @param address The address to hole fill at.
+     */
+    @Override
+    public void fillHole(long address) throws OverwriteException {
+        int numUnits = getLayout().getSegmentLength(address);
+        CompletableFuture<Boolean>[] futures = new CompletableFuture[numUnits];
+        for (int i=0; i<numUnits; i++) {
+            log.trace("fillHole[{}]: quorum {}/{}", address, i + 1, numUnits);
+            futures[i]=getLayout().getLogUnitClient(address, i).fillHole(address);
```
Comment (dahliamalkhi) [inline runtime/src/main/java/org/corfudb/runtime/view/QuorumReplicationView.java]: We must implement fill-hole in two phases (full Paxos consensus protocol). We can chat about potential optimizations, but the most straight forward way is a 2-phase consensus.

File: runtime/src/main/java/org/corfudb/runtime/view/QuorumReplicationView.java (hunk 3 above)
Comment (kspirov) [inline runtime/src/main/java/org/corfudb/runtime/view/QuorumReplicationView.java]: Good question.

File: runtime/src/main/java/org/corfudb/runtime/view/QuorumReplicationView.java (hunk 2 above)
Comment (kspirov) [inline runtime/src/main/java/org/corfudb/runtime/view/QuorumReplicationView.java]: I improved the future especially to address this, but instead of returning "unwritten" it is throwing new exception. Let me see whether it is more reasonable to return a special value.

File: runtime/src/main/java/org/corfudb/runtime/view/QuorumReplicationView.java (hunk 4 above)
Comment (kspirov) [inline runtime/src/main/java/org/corfudb/runtime/view/QuorumReplicationView.java]: Yes, I am working on this. In order to make "fill hole" testable, I created a special rule (with your framework), that "damages" the sequencer and makes it return duplicate results. It even helped me to find one small mistake in the future above. Unfortunately the rule is still "red" and has some issues with the "chain replication" too. I believe I will be able to stabilize it and update my request until our meeting on Monday.

File: runtime/src/main/java/org/corfudb/runtime/view/QuorumFutureFactory.java [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+     * ExecutionException will be thrown,  otherwise the future will return null
+     *
+     * @param futures The N futures
+     * @return The composite future
+     */
+    static <R> Future<R> getFirstWinsFuture(CompletableFuture<R>... futures) {
+        return getCompositeFuture(1, futures);
+    }
+
+    private static <R> Future<R> getCompositeFuture(int futuresToWait, CompletableFuture<R>... futures) {
```
Comment (kspirov) [inline runtime/src/main/java/org/corfudb/runtime/view/QuorumFutureFactory.java]: It should and must be thread safe, let's correct the comment.

File: runtime/src/main/java/org/corfudb/runtime/view/QuorumFutureFactory.java [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+     *
+     * @param futures The N futures
+     * @return The composite future
+     */
+    static <R> Future<R> getQuorumFuture(CompletableFuture<R>... futures) {
+        return getCompositeFuture(futures.length/2+1, futures);
+    }
+
+    /**
+     * Get a future that will complete only when a single futures complete or n/2+1 futures are canceled.
```
Comment (kspirov) [inline runtime/src/main/java/org/corfudb/runtime/view/QuorumFutureFactory.java]: Actually (as we discussed earlier in this thread) there are 3 cases:
1) Quorum is reached
2) Quorum is not reached due to different values. . We should return exception.
3) Quorum is not reached due to failure in more than one node.

In both cases 2 and 3 we should return exception - so I agree with your remark (even when canceling you should not return null, this is inconsistent). 

Also there was inconsistency from the fact you have different variables futuresToWait and quorum (calculated here). 

In the next version I am preparing, the code here is simply the following:

 boolean noMoreHope = numUncompletedFutures+greatestNumCompleteFutures < futuresToWait;
  if (noMoreHope) {
     done = canceled = true;
       ..... // throw exception

So we are avo

File: runtime/src/main/java/org/corfudb/runtime/view/QuorumFutureFactory.java [hunk 7]
```diff
@@ -0,0 +1,139 @@
+/**
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.corfudb.runtime.view;
+
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.concurrent.*;
```
Comment (kspirov) [inline runtime/src/main/java/org/corfudb/runtime/view/QuorumFutureFactory.java]: Ok

File: runtime/src/main/java/org/corfudb/runtime/view/QuorumFutureFactory.java [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+     * @param futures The N futures
+     * @return The composite future
+     */
+    static <R> Future<R> getFirstWinsFuture(CompletableFuture<R>... futures) {
+        return getCompositeFuture(1, futures);
+    }
+
+    private static <R> Future<R> getCompositeFuture(int futuresToWait, CompletableFuture<R>... futures) {
+        return new Future<R>() {
+            private boolean done = false;
```
Comment (kspirov) [inline runtime/src/main/java/org/corfudb/runtime/view/QuorumFutureFactory.java]: Thanks for this find, these two variables should be volatie.

File: runtime/src/main/java/org/corfudb/runtime/view/AbstractReplicationView.java [hunk 9]
```diff
@@ -40,20 +40,7 @@ public AbstractReplicationView(Layout layout, Layout.LayoutSegment ls) {
...
-                    log.warn("Quorum replication is not yet supported!");
-                    break;
-                case REPLEX:
-                    return new ReplexReplicationView(l, ls);
-                default:
-                    log.error("Unknown replication mode {} selected.", mode);
-                    break;
-            }
-            return null;
+            return mode.getReplicationView(l, ls);
```
Comment (kspirov) [inline runtime/src/main/java/org/corfudb/runtime/view/AbstractReplicationView.java]: Dalia, Michael?

Comment (dahliamalkhi) [inline runtime/src/main/java/org/corfudb/runtime/view/AbstractReplicationView.java]: In theory, you can change the replication mode with a layout change, and every segment in the global sequence of operations is independent from others.

In practice, I don't envision this being terribly useful.

**Comment by kspirov** [comment]
Read quorum - only if quorum from the same values is reached, changes… in the future from the review request

This change is just about the address the reviews from  Dalia, Kevin and Medhavi, NOT about recovery-write/2 step hole fill, that hopefully will come next (later tonight?).

The change corrects the future and allows conflicting results in it. The reason - when using this future to read (https://github.com/CorfuDB/CorfuDB/wiki/Quorum-Replication) we need a quorum of the SAME values, not just quorum of any value. The reason is that during the step 2 of the recovery write, there is possibility of conflicting values (if a quorum is reached at step 1, then a client with a higher rank reaches a quorum, the first client might still be able to write to some of the servers).

Now in the future we counting the values, for lighter implementation we are reusing the TreeMultimap from Guava. We are comparing by value the ReadResponses by serializing them, which is not so quick, but as it is used only rarery, I don't believe it will be a problem. If there is no way to reach a decision, an exceptio is thrown.

**Comment by kspirov** [comment]
Quorum replication, code/feature complete milestone  …

Sending you the code just for review, as I want to show how I implement all aspects of the quorum replication.  If we are OK with my approach, the next submission will contain
better unit test coverage (still working on it),  minor bugfixes, extra logging, reactions to review comments, and... that's it.

So - this implementation was a real adventure.

IN ODRER TO IMPLEMENT THE SPEC OF DALIA, I DID: https://github.com/CorfuDB/CorfuDB/wiki/Quorum-Replication

1) A separate stream view (in the style of Michael) that is able to change position, and it keeps and increases the rank.

2) The replication view is able to send different hole and write messages for the different phases.

3) So we have new WriteModes enums for the WriteRequest message.

4) In the same way we have a new new FillHole mode enum in FillHlleRequest message. This is the only change, that concerns the existing chain replication mode, but a 2 new bytes for this rare message should not be a problem.

5) Two new data types are defined DATA_PROPOSED and HOLE_PROPOSED, but these 2 data types never go to the log, they are only in the in-memory cache in quorum delegate
for the log server (see below)

6) On the server side: the replex and new quorum are extracted to delegates in order not to bloat the default implementation

7) The quorum implementation keeps the DATA_PROPOSED and HOLE_PROPOSED messages in internal cache. I permit this cache to expire to prevent DoS.

8) The business logic with the comparing of the ranks from the spec is in the quorum delegate

9) A synchronization/locking is needed here, as our updates are needed on reads. But my implementation is smart enough not to look every write message, but only the message on the specified address (see PrefixBasedLocalLocks class, I believe it's  quite interesting approach. ).

REMARK: The non steady state is detected by the client in two ways
1) Write/read conflicts between the values.
2) By new method in the sequencer, which is currently just for unit testing. Making the sequencer changeable and having a real backup sequencer  as specified by Dalia is is different efort.

END OF THE EXPECTED FEATURES. LET'S HAVE A STEP BACK:

WHAT WAS NOT IN THE DOCUMENT OF DALIA, BUT I NEEDED TO IMPLEMENT:
 - Step3 (optional,  asynchronous) - overriding of conflicting values  after successful write or hole fill. Luckily there are no such values, but doing this was the most difficult task.

ONE MORE STEP BACK - I realized a few days ago that I must do this in order to have a production level quality.  The issue something we already discussed, when we were working on the "Future class -
quorum algorithm gives warranty that after step 2 we will have a quorum of the same values, but  it DOES NOT give warranty, that the some different values won't be stored. So - imagine 3 servers, quorum of 2 with value A and sinle server with value B. One of the first 2 server dies, and we will have only A against B - the quorum is lost.

In order to attack this, after the write gains quorum,  a bit later after the write, the scheduler quickly checks the status of the futures. If therese is still something unresolved outside of the quorum, an override request is scheduled.

Sounds simple and quick. But... on the server side I was badly surprised - I needed to deal with the override exception (and in general with the fact the override concept was not natural).  So what I am doing - I am not changing log entities (we continue to work in append only manner), I am just trimming the previous entities by maintaining a file of blacklisted indexes.

I know the last can overlap with the efforts to implement trim, but I believe I will be able to merge this into different implementation too, if necessary.  The GC can clean these "trim index" files.

**Comment by no2chem** [comment]
@kspirov please label this PR as ready if you are okay if we merge it in.

**Comment by kspirov** [comment]
Thanks for the review, Michael and sorry this review became too big. I know you and Dahlia are quite occupied these days (Radio papers review, etc). Let’s do the following – I will add a bit more unit tests, so we are more confident about the corner cases and I then will mark it as ready. Some rebase/squash at the end might be needed too, no need to keep the intermediate versions.

File: runtime/src/main/java/org/corfudb/runtime/view/SequencerView.java [hunk 10]
```diff
@@ -1,24 +1,35 @@
...
 
+
 /**
  * Created by mwei on 12/10/15.
  */
+
 public class SequencerView extends AbstractView {
 
+    // TODO the failover sequencer should be a separate class as specified
+    // in https://github.com/CorfuDB/CorfuDB/wiki/Quorum-Replication
```
Comment (dahliamalkhi) [inline runtime/src/main/java/org/corfudb/runtime/view/SequencerView.java]: Sequencer failover is now allowed only within a full-fledged layout-change. It establishes a new layout, as well as a new epoch. Therefore, there is no need to manage sequencer leases as previously described in the wiki.

A recovery-sequencer is not yet implemented, but that's a different story. It is on my TODO list, and is relatively simple.

File: runtime/src/main/java/org/corfudb/runtime/view/SequencerView.java [hunk 11]
```diff
@@ -1,24 +1,35 @@
...
+    @Setter(AccessLevel.PACKAGE)
+    @Getter(AccessLevel.PACKAGE)
+    private boolean failover = false;
```
Comment (dahliamalkhi) [inline runtime/src/main/java/org/corfudb/runtime/view/SequencerView.java]: should not be needed.

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **e50f338**.

**Throughput**: _Average_: 35.7142857143, _Standard Deviation_: 13.0270519164
**Latency**: _Average_: 263.74, _Standard Deviation_: 131.31369464

 An interactive dashboard with Pull Request Performance Metrics is available at:
[Pull Request #438 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/Et6jlCvlRZHhbZQDZIt546HgewH8AJ6c)

File: runtime/src/main/java/org/corfudb/runtime/view/SequencerView.java (hunk 11 above)
Comment (kspirov) [inline runtime/src/main/java/org/corfudb/runtime/view/SequencerView.java]: OK

File: infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServerQuorumReplicationDelegate.java [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+                        return;
+                    }
+                    if (previous.getRank()>=msgRank) {
+                        r.sendResponse(ctx, msg, CorfuMsgType.ERROR_RANK.msg());
+                        return;
+                    }
+                    localCache.put(addr, msg.getPayload().getData()); // for phase 1 - write only to the memory cache
+                    return;
+                } else { // Phase 2 or regular write
+                    if (previous.getRank()>msgRank || (previous.getRank()>msgRank && !phase1Matches)) {
```
Comment (kspirov) [inline infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServerQuorumReplicationDelegate.java]: yes, it is a bug, it needed to be

if (previous.getRank()>msgRank || (previous.getRank()==msgRank && !phase1Matches)) {

File: runtime/src/main/java/org/corfudb/runtime/view/stream/QuorumStreamView.java [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+                // to inform the client we didn't get the address.
+                if (deacquisitionCallback != null) {
+                    if (!deacquisitionCallback.apply(tokenResponse)) {
+                        log.debug("Deacquisition requested abort");
+                        return -1L;
+                    }
+                }
+                // Request a new token, informing the sequencer we were
+                // overwritten.
+                tokenResponse = runtime.getSequencerView()
```
Comment (dahliamalkhi) [inline runtime/src/main/java/org/corfudb/runtime/view/stream/QuorumStreamView.java]: I this case, since you obtain a new token, the rank should not be incremented; it is like a normal write, using rank=0.

I think that the wiki is a bit mis-leading regarding the 1-QR "recovery" write. It is really not that different from hole-write. I think we want to leave it out of the protocol, and just implement hole-fill correctly.

**Comment by kspirov** [comment]
Closing the pull request, the thread continues in https://github.com/CorfuDB/CorfuDB/pull/449
